x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -431,14 +444,20 @@ public void authenticateToken(SecureString tokenString, ActionListener<Authentic
      */
     public void getAuthenticationAndMetadata(String token, ActionListener<Tuple<Authentication, Map<String, Object>>> listener) {
         decodeToken(token, ActionListener.wrap(
-                userToken -> {
-                    if (userToken == null) {
-                        listener.onFailure(new ElasticsearchSecurityException("supplied token is not valid"));
-                    } else {
-                        listener.onResponse(new Tuple<>(userToken.getAuthentication(), userToken.getMetadata()));
-                    }
-                },
-                listener::onFailure
+            userToken -> {
+                if (userToken == null) {
+                    listener.onFailure(new ElasticsearchSecurityException("supplied token is not valid"));
+                } else {
+                    listener.onResponse(new Tuple<>(userToken.getAuthentication(), userToken.getMetadata()));
+                }
+            },
+            e -> {
+                if (isShardNotAvailableException(e)) {
+                    listener.onResponse(null);
```

Comment:
Is this going to cause a NPE down the line, e.g. https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlLogoutAction.java#L57

Reply:
Good catch thanks, I will revert this part of the change. We should fail this either way and null doesn't make sense for this Listener